# Detect poisoned mushrooms with decision tree classifier in python
09-detect-poisoned-mushrooms-w-dt-in-python

In [34]:
# Setup libraries
import pandas as pd
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import classification_report

In [35]:
# Read data into dataframe
df = pd.read_csv('../00-Datasets/mushrooms.csv')

In [36]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8124 entries, 0 to 8123
Data columns (total 23 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   class                     8124 non-null   object
 1   cap-shape                 8124 non-null   object
 2   cap-surface               8124 non-null   object
 3   cap-color                 8124 non-null   object
 4   bruises                   8124 non-null   object
 5   odor                      8124 non-null   object
 6   gill-attachment           8124 non-null   object
 7   gill-spacing              8124 non-null   object
 8   gill-size                 8124 non-null   object
 9   gill-color                8124 non-null   object
 10  stalk-shape               8124 non-null   object
 11  stalk-root                8124 non-null   object
 12  stalk-surface-above-ring  8124 non-null   object
 13  stalk-surface-below-ring  8124 non-null   object
 14  stalk-color-above-ring  

In [37]:
df.head()

class cap-shape cap-surface cap-color bruises odor gill-attachment  \
0     p         x           s         n       t    p               f   
1     e         x           s         y       t    a               f   
2     e         b           s         w       t    l               f   
3     p         x           y         w       t    p               f   
4     e         x           s         g       f    n               f   

  gill-spacing gill-size gill-color  ... stalk-surface-below-ring  \
0            c         n          k  ...                        s   
1            c         b          k  ...                        s   
2            c         b          n  ...                        s   
3            c         n          n  ...                        s   
4            w         b          k  ...                        s   

  stalk-color-above-ring stalk-color-below-ring veil-type veil-color  \
0                      w                      w         p          w   
1                      w                      w         p          w   
2                      w                      w         p          w   
3                      w                      w         p          w   
4                      w                      w         p          w   

  ring-number ring-type spore-print-color population habitat  
0           o         p                 k          s       u  
1           o         p                 n          n       g  
2           o         p                 n          n       m  
3           o         p                 k          s       u  
4           o         e                 n          a       g  

[5 rows x 23 columns]

In [38]:
df.tail()

class cap-shape cap-surface cap-color bruises odor gill-attachment  \
8119     e         k           s         n       f    n               a   
8120     e         x           s         n       f    n               a   
8121     e         f           s         n       f    n               a   
8122     p         k           y         n       f    y               f   
8123     e         x           s         n       f    n               a   

     gill-spacing gill-size gill-color  ... stalk-surface-below-ring  \
8119            c         b          y  ...                        s   
8120            c         b          y  ...                        s   
8121            c         b          n  ...                        s   
8122            c         n          b  ...                        k   
8123            c         b          y  ...                        s   

     stalk-color-above-ring stalk-color-below-ring veil-type veil-color  \
8119                      o                      o         p          o   
8120                      o                      o         p          n   
8121                      o                      o         p          o   
8122                      w                      w         p          w   
8123                      o                      o         p          o   

     ring-number ring-type spore-print-color population habitat  
8119           o         p                 b          c       l  
8120           o         p                 b          v       l  
8121           o         p                 b          c       l  
8122           o         e                 w          v       l  
8123           o         p                 o          c       l  

[5 rows x 23 columns]

In [39]:
df.sample(6)

class cap-shape cap-surface cap-color bruises odor gill-attachment  \
6888     p         x           y         e       f    s               f   
2604     e         f           f         n       t    n               f   
98       e         b           s         w       t    a               f   
5612     e         f           s         n       t    n               f   
4106     p         b           s         b       t    n               f   
1609     e         x           s         n       f    n               f   

     gill-spacing gill-size gill-color  ... stalk-surface-below-ring  \
6888            c         n          b  ...                        s   
2604            c         b          n  ...                        s   
98              c         b          g  ...                        s   
5612            c         b          w  ...                        s   
4106            c         b          g  ...                        s   
1609            w         b          k  ...                        f   

     stalk-color-above-ring stalk-color-below-ring veil-type veil-color  \
6888                      p                      p         p          w   
2604                      g                      w         p          w   
98                        w                      w         p          w   
5612                      w                      e         p          w   
4106                      w                      w         p          w   
1609                      w                      w         p          w   

     ring-number ring-type spore-print-color population habitat  
6888           o         e                 w          v       p  
2604           o         p                 n          v       d  
98             o         p                 n          s       g  
5612           t         e                 w          c       w  
4106           t         p                 r          v       m  
1609           o         e                 n          s       g  

[6 rows x 23 columns]

In [40]:
# Check for missing values
df.isna().sum()

class                       0
cap-shape                   0
cap-surface                 0
cap-color                   0
bruises                     0
odor                        0
gill-attachment             0
gill-spacing                0
gill-size                   0
gill-color                  0
stalk-shape                 0
stalk-root                  0
stalk-surface-above-ring    0
stalk-surface-below-ring    0
stalk-color-above-ring      0
stalk-color-below-ring      0
veil-type                   0
veil-color                  0
ring-number                 0
ring-type                   0
spore-print-color           0
population                  0
habitat                     0
dtype: int64

In [41]:
# Check for duplicates
df.duplicated().sum()

0

In [42]:
# Explore some features
df['class'].value_counts()

e    4208
p    3916
Name: class, dtype: int64

In [43]:
# Explore tabular summary
df.describe()

class cap-shape cap-surface cap-color bruises  odor gill-attachment  \
count   8124      8124        8124      8124    8124  8124            8124   
unique     2         6           4        10       2     9               2   
top        e         x           y         n       f     n               f   
freq    4208      3656        3244      2284    4748  3528            7914   

       gill-spacing gill-size gill-color  ... stalk-surface-below-ring  \
count          8124      8124       8124  ...                     8124   
unique            2         2         12  ...                        4   
top               c         b          b  ...                        s   
freq           6812      5612       1728  ...                     4936   

       stalk-color-above-ring stalk-color-below-ring veil-type veil-color  \
count                    8124                   8124      8124       8124   
unique                      9                      9         1          4   
top                         w                      w         p          w   
freq                     4464                   4384      8124       7924   

       ring-number ring-type spore-print-color population habitat  
count         8124      8124              8124       8124    8124  
unique           3         5                 9          6       7  
top              o         p                 w          v       d  
freq          7488      3968              2388       4040    3148  

[4 rows x 23 columns]

In [44]:
# Prepare the data
le = LabelEncoder()
for col in df.columns:
    df[col] = le.fit_transform(df[col])  
# df['class'] = le.fit_transform(df['class'])

In [45]:
# Split dataset into 80% train and 20% test
X = df.drop('class', axis=1)
y = df['class']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

In [46]:
# Build model
model = DecisionTreeClassifier()
model.fit(X_train, y_train)

DecisionTreeClassifier()

In [47]:
# Feed test data into model
y_predict = model.predict(X_test)

actual_vs_predict = pd.DataFrame({'Actual': y_test,
                                 'Prediction': y_predict})
actual_vs_predict.sample(12)

Actual  Prediction
7145       1           1
5971       1           1
3990       1           1
6997       0           0
6660       0           0
2608       0           0
4539       1           1
1563       0           0
3998       0           0
6217       1           1
2530       0           0
4559       1           1

In [48]:
# Evaluate model
print(classification_report(y_test, y_predict))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00       852
           1       1.00      1.00      1.00       773

    accuracy                           1.00      1625
   macro avg       1.00      1.00      1.00      1625
weighted avg       1.00      1.00      1.00      1625



## Notes
* We wanted to detect poisoned mushrooms.
* Our dataset had 8124 records and 23 features, all of which are of data type object.
* A quick look showed that there were no missing or duplicated values.
* The dataset was split into 80% train and 20% test sets. No validation set was created or used. No hyperparameter tuning occurred.
* We built, trained, and tested a decision tree classifier.
* We evaluated our model using ground truth and several metrics including precision, recall, and f1-score. 